### DISCUSSION TITLE
Request for Technical Support: Database Compaction

### DISCUSSION THREAD

**Discussion by fernandox2000** [ravendb/ravendb#18314] [Q&A]
Hi all,

I'm are currently developing a feature that involves calling the Compact Database Operation from RavenDB. The operation itself is functioning as expected and successfully compacts the database. However, I am encountering an issue when attempting to test this method.
I'm utilizing the Raven TestDriver to create the database in memory. However, when we attempt to call the compaction operation, we encounter an exception. Here is a snippet of the code along with the exception we are facing:
 
![image](https://github.com/ravendb/ravendb/assets/70704663/e4b6a066-f7ee-4a04-83fb-205aaea6906c)

![image](https://github.com/ravendb/ravendb/assets/70704663/547faa1d-77df-48bb-a5ca-6e8cba0077a5)

Exception Message:
System.InvalidOperationException: Failed to execute compaction for CompactDatabaseExceptionTest
---> System.InvalidCastException: Unable to cast object of type 'PureMemoryStorageEnvironmentOptions' to type 'DirectoryStorageEnvironmentOptions'.

It's not possible to mock the method, as it is a concrete part of Raven. Additionally, testing it in the in-memory database isn't working, as I imagine it's not possible to use the compact operation in the in-memory database.

What are the recommended steps to proceed with testing the compact operation, considering the current limitations I've encountered with the TestDriver? Are there alternative approaches or adjustments I can make to successfully conduct this test? am I doing something wrong?
I already thought on creating a local temporary database for using the compact for the test, but It seems I need to use the TestDriver, I don't know if it is possible with it.

**Comment by ayende** [comment]
As you noted, it is not possible to run compaction on in memory database. This is because it doesn't really make sense to do so.

If you are interested in actually testing compaction, you need to do that on a real database. 
As you are using RavenDB TestDriver, by _default_ we are using in memory databases. But you can override that setting.

Here is how we define this setting:https://github.com/ravendb/ravendb/blob/9d59f0ff95bacef9d2c6a2df9c2a7c39523768fe/src/Raven.TestDriver/RavenTestDriver.cs#L282

As you can see, it applies globally to all databases in the system.

You can inherit from `RavenTestDriver` and override the `PreConfigureDatabase` method, something like this:


```
protected overridevoid PreConfigureDatabase(DatabaseRecord databaseRecord)
{
    databaseRecord.Settings["RunInMemory"] = "false";
}
```

I would recommend that you do this only for that specific test, of course.
